Remove redundant checksum calculation during vertical merge of parts with projections

amosbird opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Touches MergeTree part checksum generation around projections, which is integrity-critical; a missed or duplicated checksum could surface as corrupted parts during mutation/merge workflows.
> 
> **Overview**
> **Removes redundant projection checksum bookkeeping during column-only writes.** `MergedColumnOnlyOutputStream::fillChecksums` no longer adds projection-part entries to the part checksums, avoiding duplicate work during vertical merges where projections are accounted for later.
> 
> **Preserves mutation correctness by adding projection checksums at mutation finalize time.** `MutateTask` now explicitly appends checksums for rebuilt projection parts after `fillChecksums`, ensuring mutations that rebuild projections still persist the required `*.proj` checksum records.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit debe999c65acf2ad08bf4460c10e5ae032226c86. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

<!-- ch-version-info:start -->
### Version info
- Merged into: `26.6.1.491`
<!-- ch-version-info:end -->

amosbird committed:
Remove unnecessary checksum operation

tiandiwonder commented on src/Storages/MergeTree/MergedColumnOnlyOutputStream.cpp:
```diff
@@ -93,12 +93,6 @@ MergedColumnOnlyOutputStream::fillChecksums(
     for (const auto & filename : checksums_to_remove)
         all_checksums.files.erase(filename);
 
-    for (const auto & [projection_name, projection_part] : new_part->getProjectionParts())
```
Are there any situations we need to recalculate checksum for projections? such as projection column has been chagned or not full part storage(s3)?

alexey-milovidov replied:
The projection checksums are still added, just from different call sites. `MergedColumnOnlyOutputStream::fillChecksums` has two callers:

1. **Vertical merge** — `MergeTask::VerticalMergeStage::finalizeVerticalMergeForOneColumn` calls it per gathering column; the result accumulates in `gathered_data.checksums`. The part is then finalized by `MergedBlockOutputStream::finalizePartAsync`, which itself adds projection checksums (see `MergedBlockOutputStream.cpp` around line 220). So the addition that was removed here was redundant — `finalizePartAsync` overwrote it with the same value.
2. **Mutations** — `MutateSomePartColumnsTask::finalize` has no later `finalizePartAsync`, so this PR's second commit adds the projection checksum step there explicitly.

The recalculation of projection part *contents* (when a column a projection depends on changes) happens earlier: `projections_to_recalc` collects affected projections, `PartMergerWriter` rebuilds them, and the rebuilt projection parts end up in `new_data_part->getProjectionParts()` with their final checksums. So when the new code iterates `getProjectionParts()` and calls `getTotalChecksumUInt128`, it sees up-to-date checksums — and storage type (full part / S3 / etc.) does not change this path, since the iteration reads in-memory data populated when each projection part was finalized.

tiandiwonder reviewed (commented):
fix test failure: 01710_projection_fetch_long

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into remove-useless-code-in-projection-1

# Conflicts:
#	src/Storages/MergeTree/MergedColumnOnlyOutputStream.cpp

alexey-milovidov committed:
Add projection checksums in MutateSomePartColumnsTask::finalize

The previous commit removed the projection-checksums loop from
`MergedColumnOnlyOutputStream::fillChecksums`. That removal is correct
for the vertical merge path (where `MergedBlockOutputStream::finalizePartAsync`
re-adds projection checksums later), but for `MutateSomePartColumnsTask`
there is no subsequent step that adds them, so the resulting
`checksums.txt` was missing `*.proj` entries for projections rebuilt
during a mutation (e.g. `MATERIALIZE PROJECTION`).

This caused stateless test `01710_projection_fetch_long` to fail because
the replicated part's checksums no longer contained the projection entry
after `materialize projection`, and replicas observed inconsistent state.

Restore the loop in the mutation finalize path so that projection
checksums end up in `new_data_part->checksums` before
`finalizeMutatedPart` writes them to disk.

CI: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=92866&sha=0e81c92c1f24f056d3672f0e1f34603dffb55596&name_0=PR
PR: https://github.com/ClickHouse/ClickHouse/pull/92866

alexey-milovidov commented:
@groeneai, provide a fix for `FileCacheTest`.

groeneai commented:
@alexey-milovidov @amosbird — `FileCacheTest.SLRUFreeSpaceKeepingProtectedOnly` is **not caused by this PR**, it's a chronic flake hitting many unrelated PRs.

### Cross-PR data (CIDB, last 30 days)

The same test fails on **21 distinct PRs** plus **3 master commits** in just 2 days:

| Where | Hits | Build |
|---|---|---|
| Master (`74e1c5ed`, `407d6cca`, `6d2c8bd3`) | 3 | `Unit tests (amd_llvm_coverage)` |
| 20 unrelated PRs (this one + 19 others) | 20 | mostly `amd_llvm_coverage`, also one `asan_ubsan`, one `tsan` |

This PR (#92866) modifies only `MergedColumnOnlyOutputStream.cpp` and `MutateTask.cpp` — neither touches `FileCache` code paths. The test failure is a timing race on slow builds, not a regression here.

### Root cause

The test was added in #104313 to cover #104307. It drives the bug condition through `FileCache::freeSpaceRatioKeepingThreadFunc` with `keep_free_space_elements_ratio = 1.0`, which causes `cache->initialize()` at `src/Interpreters/FileCache/FileCache.cpp:448` to schedule the asynchronous background eviction task. That task races with the test's `read_and_check` and intermediate `assertProbationary` / `assertProtected` calls. On slow builds (notably `Unit tests (amd_llvm_coverage)`) the background thread evicts entries between the populate and assert steps, so `getFileSegmentsNum()` reports `3` instead of the expected `0` at line 1626.

### Fix is already in flight

PR #104418 ("Fix flaky FileCacheTest.SLRUFreeSpaceKeepingProtectedOnly", open since 2026-05-08) replaces the racy integration test with a focused unit-level test that calls `SLRUFileCachePriority::collectEvictionInfo` directly with three protected-only entries via `addForRestore(... QueueEntryType::SLRU_Protected ...)`. That reproduces the exact `chassert` precondition (probationary empty, protected populated, `is_total_space_cleanup=true`) without involving `FileCache`'s asynchronous machinery, making it deterministic on every build flavor.

@kssenii was already cc'd as the filesystem-cache owner. PR is `MERGEABLE` and currently green except for an unrelated `Fast test` flake on `02932_refreshable_materialized_views_2` (separate chronic timeout, 11 failures / 10 PRs in 30d, not introduced by #104418).

### TL;DR for #92866

Re-running `Unit tests (amd_llvm_coverage)` here will pass intermittently. The test will become deterministic once #104418 merges. No action needed from @amosbird.

CI report for the failure on this PR:
https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=92866&sha=b82648604f3e517985d455be11e7939bfe265704&name_0=PR&name_1=Unit%20tests%20%28amd_llvm_coverage%29

(session: cron:clickhouse-ci-task-worker:20260510-004500)

alexey-milovidov committed:
Merge branch 'master' into remove-useless-code-in-projection-1

alexey-milovidov commented:
CI status note for the two non-success checks on commit `fccb42d`:

- **AST fuzzer (amd_debug, targeted, old_compatibility)** — `Inconsistent AST formatting (STID: 1941-1bfa)`. This is an existing fuzzer-found bug unrelated to this PR (the failing query is `ALTER TABLE ... ADD PROJECTION ... GROUP BY ... AS alias26)` where the alias placement inside a function call mis-prints). It is being fixed in #105091.
- **Code Review** — the copilot subprocess exited non-OK (`Copilot review failed after 3 attempts`); this is a transient infra failure and is what currently makes "Mergeable Check" fail.

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into remove-useless-code-in-projection-1

alexey-milovidov commented:
Merged master into the branch to re-run CI. The two failures look unrelated to this PR:

1. **`Code Review`** — Copilot review subprocess infra failure (`ERROR: Copilot review failed after 3 attempts: copilot subprocess exited with non-OK status`).
2. **`AST fuzzer (amd_debug, targeted, old_compatibility)` — `Logical error: 'Inconsistent AST formatting: the query: ...'`** — a roundtrip formatting issue for an `ADD PROJECTION` expression containing `<=>` (`isNotDistinctFrom`) with an `AS alias`. The same error reproduces on master and several unrelated PRs (queried from `play.clickhouse.com`):

```
pull_request_number  date
0                    2026-05-12   ← master
92866                2026-05-12
103234               2026-05-11
90043                2026-05-10
92340                2026-04-27
```

The PR itself only touches projection-checksum bookkeeping in vertical merge / mutation, with no parser/formatter changes — it cannot affect AST roundtrip behavior.

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into remove-useless-code-in-projection-1

alexey-milovidov committed:
Restore workflow file pinned to pre-merge state for fork-push

alexey-milovidov committed:
Restore unrelated workflow files from master

Workflow files in this branch were stale or had been deleted as a
workaround for GitHub's workflow-scope check. Restoring them to match
`master` so the PR diff no longer contains those unrelated changes.

alexey-milovidov committed:
Merge master to refresh merge base

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into remove-useless-code-in-projection-1

tiandiwonder commented on src/Storages/MergeTree/MutateTask.cpp:
```diff
@@ -2461,6 +2461,20 @@ class MutateSomePartColumnsTask : public IExecutableTask
+            /// `MergedColumnOnlyOutputStream::fillChecksums` no longer adds them because that addition
+            /// is redundant during vertical merge (`MergedBlockOutputStream::finalizePartAsync` adds
+            /// them later). For mutations there is no such later step, so add them here.
+            for (const auto & [projection_name, projection_part] : ctx->new_data_part->getProjectionParts())
```
why check empty here?

alexey-milovidov replied:
Unlike the vertical-merge path (where `fillChecksums` used to add these, and `new_part->getProjectionParts()` only holds freshly merged projection parts that always have real content), in the mutation path `ctx->new_data_part->getProjectionParts()` is populated by `loadProjections` (called earlier in this task). `loadProjections` can insert a projection part as a *placeholder* with empty checksums — when the projection fails to load (`IMergeTreeDataPart.cpp`, the `catch` that calls `setBrokenReason`) or when its directory is missing while consistency checking. Such a part is marked broken and its `checksums` are default-constructed/empty.

Without the `empty()` guard we would write a `.proj` entry into `checksums.txt` with `getTotalSizeOnDisk() == 0` and a zero `getTotalChecksumUInt128()` for a projection that has no real on-disk data, which would later fail validation. The guard skips those and only records checksums for projections that were actually built/rebuilt with content.

tiandiwonder replied:
Confirmed the ordering empirically on a local instrumented build. One detail worth refining: in this path `loadProjections` runs *after* this loop, not before it, so it isn't the source of the entries this loop iterates.

To exercise `MutateSomePartColumnsTask` you need a **Wide** part — a Compact part rewrites all columns even for a single-column `UPDATE`, so it routes to `MutateAllPartColumnsTask` (which never reaches this loop). With a Wide part, a table with two projections where `UPDATE b` rebuilds only `p1` (depends on `b`) and leaves `p2` (depends on `a`, `c`) inherited via hardlink:

| Marker | `getProjectionParts()` |
|---|---|
| in `finalize`, at the `.proj` loop site | size=1 — only `p1` (rebuilt), `checksums.empty()=false` |
| `finalizeMutatedPart`, before `loadProjections` | size=1 — only `p1` |
| `finalizeMutatedPart`, after `loadProjections` | size=2 — `p1` + `p2`, both `checksums.empty()=false` |

So the inherited `p2` appears only *after* `loadProjections`, which is invoked from `finalizeMutatedPart` — i.e. after this loop has already run. At the loop site, `getProjectionParts()` holds only the projection parts rebuilt by `PartMergerWriter`; those are attached after `MergeTreeTemporaryPart::finalize`, so they always carry non-empty checksums.

Net: on the normal mutation path the `empty()` guard is unreachable — it's a defensive guard, and harmless to keep (skipping a zero-checksum entry is safe). The accurate justification is "freshly rebuilt projection parts always have checksums", rather than `loadProjections` inserting empty-checksum placeholders that this loop would see — those placeholders are only loaded later, after this loop.

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into remove-useless-code-in-projection-1

alexey-milovidov commented:
Merged the latest `master` (the branch was ~1400 commits behind). The merge was clean — only `MutateTask.cpp` auto-merged, and the PR's two changes are intact. I re-verified the correctness invariant on current master: the vertical-merge path still adds the projection checksums via `MergedBlockOutputStream::finalizePartAsync` (`MergedBlockOutputStream.cpp:222`), so removing the duplicate addition in `MergedColumnOnlyOutputStream::fillChecksums` remains correct, and the mutation path is covered by the new loop in `MutateSomePartColumnsTask::finalize`. Both changed files compile cleanly against current master. CI was green before the merge; all review threads are resolved.

alexey-milovidov commented:
The only red check was \`CH Inc sync\` (private sync PR [#44919](https://github.com/ClickHouse/clickhouse-private/pull/44919)), failing on a single integration test: \`test_shared_merge_tree_backups/test.py::test_restore_multiple_tables_on_parallel_nodes\` (`Data does not matched after BACKUP/RESTORE`).

This failure is unrelated to this PR and flaky:
- The test tables are \`SharedMergeTree() ORDER BY tuple()\` with no projections, so the changed code (`MergedColumnOnlyOutputStream::fillChecksums` and `MutateTask`, both iterating `getProjectionParts()`) is a no-op for them. The test exercises `BACKUP`/`RESTORE`, not mutations or vertical merges.
- The test passed on the in-job retry on the same binary (`PASSED` on `gw0`); a checksum regression would fail deterministically. `Data does not matched after BACKUP/RESTORE` on parallel-node restore is a known backup/restore replication race.

I re-triggered the failed private CI jobs ([run 27053851237](https://github.com/ClickHouse/clickhouse-private/actions/runs/27053851237)); the check should go green. The branch is approved, all review threads are resolved, and it merges cleanly with `master`.
